pydatastructs/strings/suffix_tree.py

```diff
@@ -0,0 +1,144 @@
+                    # prefix is already in tree
+                    break
+                parent_node = self._split_edge(e, self.active)
+
+            self.nodes.append(SuffixTreeNode())
+            e = SuffixTreeEdge(last_char_index, self.N, parent_node, len(self.nodes) - 1)
+            self._insert_edge(e)
+
+            if last_parent_node > 0:
+                self.nodes[last_parent_node].suffix_node = parent_node
+            last_parent_node = parent_node
+
+            if self.active.source_node_index == 0:
+                self.active.first_char_index += 1
+            else:
+                self.active.source_node_index = self.nodes[self.active.source_node_index].suffix_node
+            self._canonize_suffix(self.active)
+        if last_parent_node > 0:
+            self.nodes[last_parent_node].suffix_node = parent_node
+        self.active.last_char_index += 1
+        self._canonize_suffix(self.active)
+
+    def _insert_edge(self, edge):
+        self.edges[(edge.source_node_index,
+                    self.string[edge.first_char_index])] = edge
+
+    def _remove_edge(self, edge):
+        self.edges.pop(
+            (edge.source_node_index, self.string[edge.first_char_index]))
+
+    def _split_edge(self, edge, suffix):
+        self.nodes.append(SuffixTreeNode())
+        e = SuffixTreeEdge(edge.first_char_index, edge.first_char_index + suffix.length, suffix.source_node_index,
+                 len(self.nodes) - 1)
+        self._remove_edge(edge)
+        self._insert_edge(e)
+        # need to add node for each edge
+        self.nodes[e.dest_node_index].suffix_node = suffix.source_node_index
+        edge.first_char_index += suffix.length + 1
+        edge.source_node_index = e.dest_node_index
+        self._insert_edge(edge)
+        return e.dest_node_index
+
+    def _canonize_suffix(self, suffix):
+
+        if not suffix.explicit():
+            e = self.edges[suffix.source_node_index,
+                           self.string[suffix.first_char_index]]
+            if e.length <= suffix.length:
+                suffix.first_char_index += e.length + 1
+                suffix.source_node_index = e.dest_node_index
+                self._canonize_suffix(suffix)
+
+    # Public methods
+    def find_substring(self, substring):
+
+        if not substring:
+            return -1
+        if self.case_insensitive:
+            substring = substring.lower()
+        curr_node = 0
+        i = 0
+        while i < len(substring):
+            edge = self.edges.get((curr_node, substring[i]))
+            if not edge:
+                return -1
+            ln = min(edge.length + 1, len(substring) - i)
+            if substring[i:i + ln] != self.string[edge.first_char_index:edge.first_char_index + ln]:
+                return -1
+            i += edge.length + 1
+            curr_node = edge.dest_node_index
+        return edge.first_char_index - len(substring) + ln
+
+    def has_substring(self, substring):
```

Comment:
```suggestion
    def has(self, substring):
```

Add docstring for it following the same pattern as in https://github.com/codezonediitj/pydatastructs/blob/2482adb276055403485266d6d02a1489451c8e76/pydatastructs/trees/binary_trees.py#L81-L98

Notice that we specify one line on what the function is doing, then Parameters section and then Returns section.